misc/inmanta-workon-register.sh

```diff
@@ -0,0 +1,209 @@
+    )
+
+    if [ ! "$?" -eq 0 ]; then
+        echo "ERROR: Failed to determine server state directory. Is the server config valid?" >&2
+        return 1
+    fi
+    if [ ! -d "$result" ]; then
+        # only warn, don't return: path might still be valid
+        echo "WARNING: no environments directory found at '$result'. This is expected if no environments have been compiled yet. Otherwise, make sure you use this function on the server host." >&2
+    fi
+
+    echo "$result"
+    return 0
+}
+
+
+function __inmanta_workon_cli_port {
+    # Writes the configured server port to stdout
+
+    declare result
+    result=$(
+        "$INMANTA_WORKON_PYTHON" -c 'from inmanta.server.config import get_bind_port; print(get_bind_port());' 2> /dev/null
+    )
+
+    if [ ! "$?" -eq 0 ]; then
+        echo "ERROR: Failed to determine server bind port. Is the server config valid?" >&2
+        return 1
+    fi
+
+    echo "$result"
+    return 0
+}
+
+
+function __inmanta_workon_cli {
+    # Calls inmanta-cli with appropriate host and port options
+
+    declare port
+    port="$(__inmanta_workon_cli_port)" || return  # propagate error
+    "$INMANTA_WORKON_CLI" --host localhost --port "$port" "$@" 2> /dev/null
+}
+
+
+function __inmanta_workon_list {
+    # Writes a list of environments to stdout. Attempts to print a nice table if inmanta-cli works, otherwise falls back to a plain uuid list.
+
+    __inmanta_workon_cli environment list
+    if [ ! "$?" -eq 0 ]; then
+        echo "WARNING: Failed to connect through inmanta-cli, falling back to file-based environment discovery." >&2
+        declare envs_dir
+        envs_dir="$(__inmanta_workon_environments_dir)" || return  # propagate error
+        command ls -1 "$envs_dir" 2> /dev/null
+    fi
+    return 0
+}
+
+
+function __inmanta_workon_activate {
+    # Activates the environment's venv and registers the deactivate function
+
+    declare env_name="$1"
+    declare env_id="$2"
+    declare envs_dir="$3"
+
+    if [ ! -d "$envs_dir/$env_id" ]; then
+        echo "ERROR: Directory '$envs_dir/$env_id' does not exist. This may mean the environment has never started a compile." >&2
+        return 1
+    fi
+
+    # change directory
+    command cd "$envs_dir/$env_id"
+
+    activate="$envs_dir/$env_id/.env/bin/activate"
+    if [ ! -f "$activate" ]; then
+        echo "ERROR: Environment '$envs_dir/$env_id' does not contain a venv. This may mean it has never started a compile." >&2
+        return 1
+    fi
+
+    # if we are in an active venv, deactivate it first: restores PS1 and triggers our custom deactivate logic if it's an inmanta venv
+    declare -F deactivate > /dev/null && deactivate
+    # store PS1 before sourcing activate because we don't care about virtualenv's modifications
+    declare OLD_PS1="$PS1"
+    source "$activate"
+    export PS1="($env_name) $OLD_PS1"
+
+    eval 'inmanta () { python3 -m inmanta.app; }' # workaround for #4259
```

Comment:
Don't you have to forwards the parameters of the function to the `python3 -m inmanta.app` command?

Reply:
I would think so. I started out with an alias, where that isn't required. Then I moved to the function and forgot to change it. The strange thing is that it actually still works, as you can see by defining the function `function list() { ls; }`. But I must admit I don't really see how.

Reply:
(Investigating)

Reply:
Turns out I'm stupid: I was testing with `ls` rather than `list. Apparently `ls` is so ingrained in my fingers that I didn't even notice. Concluding, you're right, I need to change this. I do wonder how the tests didn't fail on this though.

Reply:
Fixed